subsys/caf/modules/sensor_event_aggregator.c

```diff
@@ -0,0 +1,233 @@
+				DT_INST_PROP(i, buf_data_length))			\
+	static struct aggregator_buffer agg_ ## i ## _bufs[] = {			\
+	UTIL_LISTIFY(DT_INST_PROP(i, number_of_buf), __INITIALIZE_BUF, i,		\
+				DT_INST_PROP(i, buf_data_length))			\
+	};
+
+#define __DEFINE_AGGREGATOR(i)								\
+	[i].descr = DT_INST_PROP(i, descr),						\
+	[i].sensor_data_size = DT_INST_PROP(i, sensor_data_size),			\
+	[i].number_of_buf = DT_INST_PROP(i, number_of_buf),				\
+	[i].agg_buffers = (struct aggregator_buffer *) &agg_ ## i ## _bufs,
+
+#define __DEFINE_AGG							\
+	static struct aggregator aggregator_arr[] = {			\
+	DT_INST_FOREACH_STATUS_OKAY(__DEFINE_AGGREGATOR)		\
+	}
+
+#define DEFINE_AGGREGATORS					\
+	DT_INST_FOREACH_STATUS_OKAY(__DEFINE_BUF_DATA)		\
+	__DEFINE_AGG
+
+#define AGG_BUF_READY 0
+#define	AGG_BUF_OCCUPIED 1
+
+struct aggregator_buffer {
+	uint8_t *data;			/* Dynamic data. */
+	uint8_t data_len;		/* Size of single sensor data in bytes. */
+	bool buffer_status;		/* Buffer status. */
+	uint8_t sample_cnt;		/* Size of single sensor data in bytes. */
+
+};
+
+struct aggregator {
+	const char *descr;			/* Description of the sensor. */
+	struct aggregator_buffer *agg_buffers;	/* Buffers */
+	struct aggregator_buffer *active_buf;	/* Active buffer to which data will be placed */
+	uint8_t sensor_state;			/* Sensors state */
+	const uint8_t sensor_data_size;		/* Size of sensor data in bytes. */
+	uint8_t number_of_buf;			/* Number of buffers */
+};
+
+
+
+DEFINE_AGGREGATORS;
+#define AGG_NUM sizeof(aggregator_arr) / sizeof(struct aggregator)
+
+
+static struct aggregator *find_aggregator(const char *descr)
+{
+	for (size_t i = 0; i < AGG_NUM; i++) {
+		if (descr == aggregator_arr[i].descr) {
+			return &aggregator_arr[i];
+		}
+	}
+	return NULL;
+}
+
+static int clean_agg_buffer(struct aggregator *agg, struct aggregator_buffer *ab)
+{
+	if (!ab) {
+		return -EFAULT;
+	}
+	LOG_INF("Cleaning, sample_cnt %d", ab->sample_cnt);
+	ab->sample_cnt = 0;
+	ab->buffer_status = AGG_BUF_READY;
+	if (agg->active_buf == NULL) {
+		agg->active_buf = ab;
+	}
+	return 0;
+}
+
+static void send_agg_buff_data(struct aggregator *agg, struct aggregator_buffer *ab)
+{
+	ab->buffer_status = AGG_BUF_OCCUPIED;
+	for (size_t i = 0; i < agg->number_of_buf; i++) {
+		if (agg->agg_buffers[i].buffer_status == AGG_BUF_READY) {
+			agg->active_buf = &agg->agg_buffers[i];
+			break;
+		}
+		agg->active_buf = NULL;
+	}
+	struct sensor_aggregator_event *event = new_sensor_aggregator_event();
+
+	event->buf = ab->data;
+	event->sample_cnt = ab->sample_cnt;
+	event->sensor_status = agg->sensor_state;
+	event->descr = agg->descr;
+	EVENT_SUBMIT(event);
+}
+
+static void add_sensor_evt_to_aggregator(struct sensor_event *event, struct aggregator *agg)
+{
+	__ASSERT_NO_MSG(agg);
+
+	if (agg->active_buf) {
+		struct aggregator_buffer *ab = agg->active_buf;
+
+		LOG_INF("status: %d, sample_cnt: %d", ab->buffer_status, ab->sample_cnt);
+
+		if (ab->sample_cnt * agg->sensor_data_size +
```

Comment:
validate if incoming data size is as expected

Reply:
we could use some index to point to the location  - to avoid multiplication

if not done still use some temp variable (optimizer will be fine with what is here but it's will be easier to read)
e.g.
```
if (event->dyndata.size != agg->sensor_data_size) {
   // this could alternatively check if you have several samples of given size but I think sensor event send explicit sample count
   return -EBADMSG;
}

if (!agg->active_buf) {
   return -ENOMEM;
}

size_t pos = ab->sample_cnt * agg->sensor_data_size;
size_t avail = agg->buffer_size - pos;

if (avail < agg->sensor_data_size) {
   send_active_buffer(agg);
   agg->active_buf = find_empty_buffer(agg);
   if (!agg->active_buf) {
      return -ENOMEM;
   }
   pos = 0;
}

memcpy(agg->active_buf[pos], event->dyndata.data, event->dyndata.size);

if (avail < agg->sensor_data_size) {
   send_active_buffer(agg);
   agg->active_buf = find_empty_buffer(agg);
}

```